cfscrape/__init__.py

```diff
@@ -94,48 +113,89 @@ def solve_cf_challenge(self, resp, **original_kwargs):
             raise ValueError("Unable to parse Cloudflare anti-bots page: %s %s" % (e.message, BUG_REPORT))
 
         # Solve the Javascript challenge
-        params["jschl_answer"] = self.solve_challenge(body, domain)
+        answer, delay = self.solve_challenge(body, domain)
+        params["jschl_answer"] = answer
 
         # Requests transforms any request into a GET after a redirect,
         # so the redirect has to be handled manually here to allow for
         # performing other types of requests even as the first request.
         method = resp.request.method
         cloudflare_kwargs["allow_redirects"] = False
 
-        end_time = time.time()
-        time.sleep(self.delay - (end_time - start_time)) # Cloudflare requires a delay before solving the challenge
+        # Cloudflare requires a delay before solving the challenge
+        time.sleep(delay - (time.time() - start_time))
 
+        # Send the challenge response and handle the redirect manually
         redirect = self.request(method, submit_url, **cloudflare_kwargs)
-
-        redirect_location = urlparse(redirect.headers["Location"])
+        redirect_location = urlparse(redirect.headers['Location'])
         if not redirect_location.netloc:
-            redirect_url = "%s://%s%s" % (parsed_url.scheme, domain, redirect_location.path)
+            redirect_url = urlunparse(
+                (
+                    parsed_url.scheme,
+                    domain,
+                    redirect_location.path,
+                    redirect_location.params,
+                    redirect_location.query,
+                    redirect_location.fragment
+                )
+            )
             return self.request(method, redirect_url, **original_kwargs)
         return self.request(method, redirect.headers["Location"], **original_kwargs)
 
     def solve_challenge(self, body, domain):
         try:
-            js = re.search(r"setTimeout\(function\(\){\s+(var "
-                        "s,t,o,p,b,r,e,a,k,i,n,g,f.+?\r?\n[\s\S]+?a\.value =.+?)\r?\n", body).group(1)
+            challenge, ms = re.search(
+                r"setTimeout\(function\(\){\s*(var "
+                r"s,t,o,p,b,r,e,a,k,i,n,g,f.+?\r?\n[\s\S]+?a\.value\s*=.+?)\r?\n"
+                "(?:[^{<>]*},\s*(\d{4,}))?", body).groups()
+
+            # The challenge requires `document.getElementById` to get this content.
+            # Future proofing would require escaping newlines and double quotes
+            innerHTML = re.search(r"<div(?: [^<>]*)? id=\"cf-dn.*?\">([^<>]*)", body).group(1)
```

Comment:
It appears that cloudflare uses the old method without the additional div containing challenge code still in on some sites (or they switched back). Not checking innerHTML for none will raise an Exception.

Reply:
Indeed. I'll fix it.

Reply:
@lukele Thanks for reviewing. Have you had a chance to update your openssl yet?
`python -c 'import ssl; print(ssl.OPENSSL_VERSION)'`

Reply:
Not yet unfortunately, will probably do so later today. Have to check if other tools depend on a specific version. What I was wondering, do you recall what the difference in openssl use was between urllib and urllib3? Since urllib does not trigger the captcha on my Mac but urllib3 does.

Reply:
I'm not sure that I understand entirely. If you're using requests then you're using urllib3's pool manager: https://github.com/kennethreitz/requests/blob/master/requests/adapters.py#L163
The pool manager, depending on the scheme, will create a SSL context and wrap the socket: https://github.com/urllib3/urllib3/blob/master/src/urllib3/connection.py#L263

The difference between using urllib and urllib3 for this purpose is going to be the difference between `create_urllib3_context` and `create_default_context`: https://github.com/python/cpython/blob/3.7/Lib/urllib/request.py#L210

The standard `ssl` library is used in all of the above cases, we're just talking in terms of different SSL context configurations: https://github.com/python/cpython/blob/master/Lib/ssl.py#L379

https://github.com/enthought/Python-2.7.3/blob/master/Lib/httplib.py#L1161

All 3 of these accept the same arguments:
https://github.com/python/cpython/blob/master/Lib/ssl.py#L551
https://github.com/python/cpython/blob/master/Lib/ssl.py#L1199
https://github.com/urllib3/urllib3/blob/master/src/urllib3/util/ssl_.py#L228

My question is, do you mean that when you're using urllib and not `requests` that you don't get a CAPTCHA?

urllib should use the default cipher list that was compiled into whatever version of python that you're using:
https://github.com/python/cpython/blob/master/Lib/ssl.py#L119
https://github.com/python/cpython/blob/master/Lib/ssl.py#L101
https://github.com/python/cpython/blob/master/Lib/ssl.py#L1218

You can see that it depends on compiler preprocessor commands and what it would be if it differs from openssl's defaults:
https://github.com/python/cpython/blob/21a9ba1992775b5a833da28bfa0a9f028d1b6761/Modules/_ssl.c#L271-L289

Pay extra close to the comments on those lines :point_up:

Reply:
> urllib should use the default cipher list that was compiled into whatever version of python that you're using

Ah right, `requests` does modify the default list to deprecate older or add newer ciphers.

Thanks.

Reply:
You're welcome, please share any more thoughts that you might have. Also, please let me know whenever you do update your openssl so we'll know. openssl does try to maintain backwards compatibility and uses the semantic versioning system (Well almost and future version will follow the semantic version spec exactly). It's not something I'd worry about, it should be safe to just update it from `1.1.0` -> `1.1.1` and for the sake of security, I advise it.

Reply:
It looks like upgrading openssl on macOS, even using something like homebrew is not that easy. Might have to build my own python linked against openssl >= 1.1.1

Reply:
This doesn't work? https://medium.com/@katopz/how-to-upgrade-openssl-8d005554401

Reply:
The chance of it breaking a lot of stuff is too big a risk. Did something similar once, that didn't end well. But compile from source should not be a big problem. Or maybe I can patch my local homebrew installation. Should have results later today.

Reply:
@lukele How did you install python?
The official installer includes a later version of openssl? https://www.python.org/downloads/release/python-2716/
https://www.python.org/downloads/release/python-373/

Are you using the system's default python?

Reply:
@lukele
https://bugs.python.org/issue32947
https://github.com/python/cpython/pull/11612

Whenever that issue is resolved then the official python installer, which includes a private copy of openssl, will resolve that issue for you until then I'd make sure to include those fixes when building from source.